### PR TITLE
chore: Upgrate and audit ESLint

### PR DESCRIPTION
**Pull request by seanhealy**
## Changes

- Update ESLint base.
- Add ESLint plugin path patch.

### Added

- 

### Changed

- 

### Deprecated

- 

### Removed

- 

### Fixed

- 

### Security

- 

## Testing

---

[In Atlantis we use Github's built in pull request reviews](https://help.github.com/en/articles/about-pull-request-reviews).

Random photo of Atlantis

### CODE REVIEW

File: .github/pull_request_template.md [hunk 1]
```diff
@@ -6,7 +6,7 @@
   eg.
     fix(pencil): stop graphite breaking when too much pressure applied — Patch Release
     feat(pencil): add 'graphiteWidth' option — (Minor) Feature Release
-    perf(pencil): BREAKING CHANGE: remove graphiteWidth option — (Major) Breaking Release
+    feat(pencil): BREAKING CHANGE: remove graphiteWidth option — (Major) Breaking Release
```
Comment (darryltec) [inline .github/pull_request_template.md]: I realized that I was using this wrong yesterday haha!

File: package.json [hunk 2]
```diff
@@ -40,29 +40,24 @@
...
     "@types/react-dom": "^16.8.4",
-    "@typescript-eslint/eslint-plugin": "^1.10.2",
-    "@typescript-eslint/parser": "^1.10.2",
     "css-loader": "^3.0.0",
     "docz": "^1.2.0",
     "docz-theme-default": "^1.0.0",
-    "eslint": "^5.16.0",
-    "eslint-config-prettier": "^4.3.0",
-    "eslint-plugin-import": "^2.17.3",
+    "eslint": "^6.3.0",
```
Comment (darryltec) [inline package.json]: I'm assuming all of those things that got removed are now a part of this new `eslint` package?

File: packages/components/src/Tabs/Tabs.tsx [hunk 3]
```diff
@@ -17,7 +18,7 @@ export function Tabs({ children }: TabsProps) {
   const [activeTab, setActiveTab] = useState(0);
   const [overflowRight, setOverflowRight] = useState(false);
   const [overflowLeft, setOverflowLeft] = useState(false);
-  const tabRow = useRef<HTMLDivElement>(null);
+  const tabRow = useRef() as MutableRefObject<HTMLDivElement>;
```
Comment (darryltec) [inline packages/components/src/Tabs/Tabs.tsx]: oh man, i like the old one... but i guess that's the price of no-null

File: packages/components/src/FormField/FormField.tsx [hunk 4]
```diff
@@ -186,20 +186,19 @@ export const FormField = React.forwardRef(
...
       align && styles[align],
       errorMessage && styles.hasErrorMessage,
-      (invalid || errorMessage || hasErrors) && styles.invalid,
       disabled && styles.disabled,
       maxLength && styles.maxLength,
       {
         [styles.miniLabel]:
           (hasMiniLabel || type === "time" || type === "select") && placeholder,
+        [styles.invalid]:
+          invalid || errorMessage || hasErrorMessages(validations),
```
Comment (edmond-c) [inline packages/components/src/FormField/FormField.tsx]: I've no problem with this diff but curious is this initiated from a new linting rule?

Comment (edmond-c) [inline packages/components/src/FormField/FormField.tsx]: I've no problem with this diff but curious if this is initiated from a new linting rule?

File: packages/components/src/InputTime/InputTimeSafari.tsx [hunk 5]
```diff
@@ -20,12 +20,12 @@ export function InputTimeSafari({
   ...params
 }: InputTimeProps) {
   const inputTime = React.createRef<HTMLInputElement>();
+  const debouncedHandleChange = debounce(handleChange, 1000);
+  const changeHandler = generateEventHandler(debouncedHandleChange);
+  const blurHandler = generateEventHandler(handleChange);
```
Comment (edmond-c) [inline packages/components/src/InputTime/InputTimeSafari.tsx]: Should we pass these in as the dependencies array into `useLayoutEffect `? I believe `eslint-plugin-react-hooks` would complain if they aren't included. Are we planning to include the plugin? https://www.npmjs.com/package/eslint-plugin-react-hooks

File: packages/eslint-config/.eslintrc.js [hunk 6]
```diff
@@ -46,14 +45,45 @@ module.exports = {
...
     "@typescript-eslint/no-use-before-define": ["error", { functions: false }],
     "@typescript-eslint/explicit-function-return-type": "off",
+    "@typescript-eslint/no-unused-vars": "error",
+    curly: ["error", "multi-line", "consistent"],
+    "prefer-const": "error",
+    "dot-notation": "error",
     "sort-imports": ["error", { ignoreDeclarationSort: true }],
-  },
+    "guard-for-in": "error",
+    "max-statements": ["error", 12],
```
Comment (edmond-c) [inline packages/eslint-config/.eslintrc.js]: I believe allowing `ignoretoplevelfunctions` would give React functional components flexibility when applying this to the Jobber project. https://eslint.org/docs/rules/max-statements#ignoretoplevelfunctions

File: packages/eslint-config/.eslintrc.js [hunk 7]
```diff
@@ -46,14 +45,45 @@ module.exports = {
...
+    "no-bitwise": "error",
+    "no-caller": "error",
+    "no-eval": "error",
+    "no-shadow": "error",
+    "no-sequences": "error",
+    "no-new-func": "error",
+    "no-template-curly-in-string": "error",
+    "no-return-await": "error",
```
Comment (edmond-c) [inline packages/eslint-config/.eslintrc.js]: Worth re-visiting this one. From JetBrains: 
> Missing await for an async function call results in having a Promise where a value was expected.  When the 'Report for promises in return statements' checkbox is checked, this inspection will also suggest adding 'await' in return statements. While this is generally not necessary, it gives two main benefits. The first one is that you won't forget to add 'await' when surrounding your code with try-catch. The second one is that having explicit 'await' helps V8 runtime to provide async stack traces.

The second-one won't trigger this rule, but the first one is applicable. Let's see how it goes for day-to-day debugging/production error logs and revisit if necessary

**Comment by edmond-c** [comment]
@darryltec @seanhealy Oh didn't see the approval nvm 😂

**Pull request merged**